contracts/staking/StakingPool.sol

```diff
@@ -0,0 +1,594 @@
+    }
+
+    /**
+     * @notice Returns the final amount of reward due for a user
+     *
+     * @param user address to calculate rewards for
+     */
+    function currentRewards(address user)
+        external
+        view
+        returns (RewardOwed[] memory)
+    {
+        User memory _user = _users[user];
+        StakingPoolLib.Config memory _config = _stakingPoolConfig;
+
+        RewardOwed[] memory rewards = new RewardOwed[](
+            _config.rewardTokens.length
+        );
+
+        for (uint256 i = 0; i < _config.rewardTokens.length; i++) {
+            if (_config.rewardType == StakingPoolLib.RewardType.FLOATING) {
+                rewards[i] = RewardOwed({
+                    amount: _calculateFloatingReward(
+                        _config.rewardTokens[i].ratio,
+                        _user.depositAmount
+                    ),
+                    tokens: _config.rewardTokens[i].tokens
+                });
+            }
+            if (_config.rewardType == StakingPoolLib.RewardType.FIXED) {
+                rewards[i] = RewardOwed({
+                    amount: _user.rewardAmounts[i],
+                    tokens: _config.rewardTokens[i].tokens
+                });
+            }
+        }
+        return rewards;
+    }
+
+    function _initializeRewardTokens(
+        address benefactor,
+        StakingPoolLib.Reward[] calldata _rewardTokens
+    ) internal {
+        for (uint256 i = 0; i < _rewardTokens.length; i++) {
+            IERC20 tokens = IERC20(_rewardTokens[i].tokens);
+
+            emit InitializeRewards(address(tokens), _rewardTokens[i].maxAmount);
+
+            require(
+                tokens.allowance(benefactor, address(this)) >=
+                    _rewardTokens[i].maxAmount,
+                "StakingPool: invalid allowance"
+            );
+
+            require(
+                tokens.transferFrom(
+                    benefactor,
+                    address(this),
+                    _rewardTokens[i].maxAmount
+                ),
+                "StakingPool: fund tx failed"
+            );
+        }
+    }
+
+    function _withdrawWithoutRewards() internal {
+        User memory user = _users[_msgSender()];
+        require(user.depositAmount > 0, "StakingPool: not eligible");
+
+        delete _users[_msgSender()];
+        StakingPoolLib.Config memory _config = _stakingPoolConfig;
+        _transferStake(
+            uint256((user.depositAmount)),
+            IERC20(_config.stakeToken)
+        );
+    }
+
+    function _setRewardsAvailableTimestamp(uint32 timestamp) internal {
+        require(!_isStakingPeriodComplete(), "StakePool: already finalized");
+        //slither-disable-next-line timestamp
+        require(timestamp > block.timestamp, "StakePool: future rewards");
+
+        _rewardsAvailableTimestamp = timestamp;
+        emit RewardsAvailableTimestamp(timestamp);
+    }
+
+    function _transferStake(uint256 amount, IERC20 stakeToken) internal {
```

Comment:
The pool will only ever have one stake token right? `_stakingPoolConfig.stakeToken`? Why are we passing it as an arg every time?

Reply:
We are using initializers vs constructors (because the original intention was to maintain compatibility with EIP 1167 Minimal Proxy Factories) so we cant set stake to immutable, so stakeToken was packed into `Config` struct so as to save gas, since this is storage loaded anyway. Loading here would negate that saving?